### PR TITLE
add timeout handling to QuicStream

### PR DESCRIPTION
**Pull request by wfurt**
I know this is low priority but I did it as experiment while working on #55642 to at least prevent infinite hang. 
I did not want to loose so we may turn it into deaf and wait. 
The scope and risk is small IMHO as it mimics what `NetworkStream` does. 
The behavior is covered by conformance tests that would light up on `CanTimeout`

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/Resources/Strings.resx [hunk 1]
```diff
@@ -150,5 +150,14 @@
   <data name="net_quic_writing_notallowed" xml:space="preserve">
     <value>Writing is not allowed on stream.</value>
   </data>
+  <data name="net_quic_ssl_optiopn" xml:space="preserve">
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/Resources/Strings.resx:153]: Where is this used?  It's also mispelled :)

File: src/libraries/System.Net.Quic/src/Resources/Strings.resx [hunk 2]
```diff
@@ -150,5 +150,14 @@
...
+    <value>The '{0}' is not supported by System.Net.Quic.</value>
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/Resources/Strings.resx:154]:
```suggestion
    <value>'{0}' is not supported by System.Net.Quic.</value>
```

File: src/libraries/System.Net.Quic/src/Resources/Strings.resx [hunk 3]
```diff
@@ -150,5 +150,14 @@
...
+  </data>
+  <data name="net_quic_timeout_use_gt_zero" xml:space="preserve">
+    <value>Timeout can be only be set to 'System.Threading.Timeout.Infinite' or a value &gt; 0.</value>
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/Resources/Strings.resx:157]:
```suggestion
    <value>Timeout can only be set to 'System.Threading.Timeout.Infinite' or positive value.</value>
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 4]
```diff
@@ -197,6 +198,47 @@ internal MsQuicStream(MsQuicConnection.State connectionState, QUIC_STREAM_OPEN_F
 
         internal override bool CanWrite => _disposed == 0 && _canWrite;
 
+        internal override bool CanTimeout => true;
+
+        private int _readTimeout = -1;
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:203]:
```suggestion
        private int _readTimeout = Timeout.Infinite;
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 5]
```diff
@@ -197,6 +198,47 @@ internal MsQuicStream(MsQuicConnection.State connectionState, QUIC_STREAM_OPEN_F
...
+                ThrowIfDisposed();
+                if (value <= 0 && value != System.Threading.Timeout.Infinite)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value), SR.net_quic_timeout_use_gt_zero);
+                }
+                _readTimeout = value;
+            }
+        }
+
+        private int _writeTimeout = -1;
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:223]:
```suggestion
        private int _writeTimeout = Timeout.Infinite;
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 6]
```diff
@@ -648,28 +689,58 @@ internal override int Read(Span<byte> buffer)
...
+                {
+                    cts = new CancellationTokenSource(_readTimeout);
+                }
+                Task<int> t = ReadAsync(new Memory<byte>(rentedBuffer, 0, buffer.Length), cts != null ? cts.Token : default).AsTask();
                 ((IAsyncResult)t).AsyncWaitHandle.WaitOne();
                 int readLength = t.GetAwaiter().GetResult();
                 rentedBuffer.AsSpan(0, readLength).CopyTo(buffer);
                 return readLength;
             }
+            catch (OperationCanceledException) when (_readTimeout > 0)
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:705]:
```suggestion
            catch (OperationCanceledException) when (cts != null && cts.IsCancellationRequested)
```
?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 7]
```diff
@@ -648,28 +689,58 @@ internal override int Read(Span<byte> buffer)
...
             ThrowIfDisposed();
             byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length);
+            CancellationTokenSource? cts = null;
             try
             {
-                Task<int> t = ReadAsync(new Memory<byte>(rentedBuffer, 0, buffer.Length)).AsTask();
+                if (_readTimeout > 0)
+                {
+                    cts = new CancellationTokenSource(_readTimeout);
+                }
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:698]: Why is this inside the try and the one in Write isn't?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 8]
```diff
@@ -648,28 +689,58 @@ internal override int Read(Span<byte> buffer)
...
+            {
+                cts = new CancellationTokenSource(_writeTimeout);
+            }
+            try
+            {
+                Task t = WriteAsync(buffer.ToArray(), cts != null ? cts.Token : default).AsTask();
+                ((IAsyncResult)t).AsyncWaitHandle.WaitOne();
+                t.GetAwaiter().GetResult();
+            }
+            catch (OperationCanceledException) when (_writeTimeout > 0)
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:734]:
```suggestion
            catch (OperationCanceledException) when (cts != null && cts.IsCancellationRequested)
```
?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Quic/src/Resources/Strings.resx (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/Resources/Strings.resx]: that is pollution from #55877. (where the spelling was corrected)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs (hunk 7 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:698]: we should return rented buffer if this fails for what ever reason. Alternatively I could move it before Shared.Rent()

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs (hunk 6 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs]: I like it as it is more specific to the Token we create. updated.

File: src/libraries/System.Net.Quic/src/Resources/Strings.resx [hunk 9]
```diff
@@ -150,5 +150,14 @@
...
   </data>
+  <data name="net_quic_timeout_use_gt_zero" xml:space="preserve">
+    <value>Timeout can only be set to 'System.Threading.Timeout.Infinite' or a value &gt; 0.</value>
+  </data>
+  <data name="net_quic_timeout" xml:space="preserve">
+    <value>Connection timed out.</value>
+  </data>
+  <data name="net_quic_ssl_option" xml:space="preserve">
+    <value>The '{0}' is not supported by System.Net.Quic.</value>
+  </data>
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/Resources/Strings.resx:161]: Where is this used?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs [hunk 10]
```diff
@@ -79,6 +79,26 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
...
 
+        public override bool CanTimeout => _provider.CanTimeout;
+
+        public override int ReadTimeout
+        {
+            get => _provider.ReadTimeout;
+            set
+            {
+                 _provider.ReadTimeout = value;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:90]:
```suggestion
            set => _provider.ReadTimeout = value;
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs [hunk 11]
```diff
@@ -79,6 +79,26 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
...
+        }
+
+        public override int WriteTimeout
+        {
+            get => _provider.WriteTimeout;
+            set
+            {
+                 _provider.WriteTimeout = value;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:99]:
```suggestion
            set => _provider.WriteTimeout = value;
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs (hunk 6 above)
Comment (geoffkizer) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs]: Shouldn't we specifically check for the token on the Exception?

```csharp
    catch (OperationCanceledException oce) when (oce.Token == cts?.Token)
```

@stephentoub?

**Comment by geoffkizer** [comment]
I think it's fine to merge this if it helps track down test issues, but let's make sure we aren't introducing any new complications. From what I can see the code seems pretty isolated, so hopefully we aren't.

**Review by CarnaViire — APPROVED** [review]
LGTM

**Pull request merged**